examples/randlanet_classification.py

```diff
@@ -0,0 +1,243 @@
+from torch import Tensor
+from torch.nn import Sequential, Linear, LeakyReLU
+from tqdm import tqdm
+import torch_geometric.transforms as T
+from torch_geometric.transforms.base_transform import BaseTransform
+from torch_geometric.datasets import ModelNet
+from torch_geometric.loader import DataLoader
+from torch_geometric.nn import MLP, global_max_pool, BatchNorm
+from torch_geometric.nn.conv import MessagePassing
+from torch_geometric.nn.pool import knn
+
+
+# Default activation, BatchNorm, and resulting MLP used by RandLaNet authors
+lrelu02_kwargs = {"negative_slope":0.2}
+
+
+bn099_kwargs = {"momentum":0.99, "eps":1e-6}
+
+
+def default_MLP(*args, **kwargs):
+    """MLP with custom activation, bn, and dropout that are always active even an last layer."""
+    kwargs["plain_last"] = kwargs.get("plain_last", False)
+    kwargs["act"] = kwargs.get("act", "LeakyReLU")
+    kwargs["act_kwargs"] = kwargs.get("act_kwargs", lrelu02_kwargs)
+    kwargs["norm_kwargs"] = kwargs.get("norm_kwargs", bn099_kwargs)
+    return MLP(*args, **kwargs)
+
+class GlobalPooling(torch.nn.Module):
+
+    def forward(self, x, pos, batch):
+        x = global_max_pool(x, batch)
+        pos = pos.new_zeros((x.size(0), 3))
+        batch = torch.arange(x.size(0), device=batch.device)
+        return x, pos, batch
+
+
+class LocalFeatureAggregation(MessagePassing):
+    """Positional encoding of points in a neighborhood."""
+
+    def __init__(self, d_out):
+        super().__init__(aggr="add", flow="target_to_source")
+        self.mlp_encoder = default_MLP([10, d_out // 2])
+        self.mlp_attention = default_MLP([d_out, d_out], bias=False, act=None, norm=None)
+        self.mlp_post_attention = default_MLP([d_out, d_out])
+
+    def forward(self, edge_indx, x, pos):
+        out = self.propagate(edge_indx, x=x, pos=pos)  # N // 4 * d_out
+        out = self.mlp_post_attention(out)  # N // 4 * d_out
+        return out
+
+    def message(self, x_j: Tensor, pos_i: Tensor, pos_j: Tensor) -> Tensor:
+        """
+        Local Spatial Encoding (locSE) and attentive pooling of features.
+
+        Args:
+            x_j (Tensor): neighboors features (K,d)
+            pos_i (Tensor): centroid position (repeated) (K,3)
+            pos_j (Tensor): neighboors positions (K,3)
+
+        returns:
+            (Tensor): locSE weighted by feature attention scores.
+
+        """
+        dist = pos_j - pos_i
+        euclidian_dist = torch.sqrt((dist * dist).sum(1, keepdim=True))
+        relative_infos = torch.cat(
+            [pos_i, pos_j, dist, euclidian_dist], dim=1
+        )  # N//4 * K, d
+        local_spatial_encoding = self.mlp_encoder(relative_infos)  # N//4 * K, d
+        local_features = torch.cat([x_j, local_spatial_encoding], dim=1)  # N//4 * K, 2d
+
+        # attention will weight the different features of x
+        attention_scores = torch.softmax(
+            self.mlp_attention(local_features), dim=-1
+        )  # N//4 * K, d_out
```

Comment:
Here you are making attention scores from the channel direction. But RandLANet should aggregate from the K points. In fact, the message function does not have any information about KNN neighbors.

Reply:
See here for a reference implementation with 3D Tensors.

https://github.com/aRI0U/RandLA-Net-pytorch/blob/057035adf587a4e377e431f54654a090e53740f2/model.py#L107-L135

The softmax is applied to the K Dim.

Reply:
Good catch! Thank you very much for this review :100: 

To react to your comment in https://github.com/torch-points3d/torch-points3d/issues/782

> Because the inputs of the message function do not have information about the KNN at all. I am wondering how to implement such spatial pooling across the KNN points.

To my understanding, the "sum" pooling operation is applied internally by the `self.propagate(...)` call. Under the hood, the indices in edge_index are used to pool the results `message(...)`. That's where param `flow` comes into action.

I reverted back `flow` to the default `source_to_target`, because I want a sum-pooling to occur on targets i.e. the centers of neighborhoods. It is somewhat counter-intuitive, but I build `edge_index` following many examples with this line : 
```
row, col = knn(pos, pos[idx], self.num_neighbors, batch_x=batch, batch_y=batch[idx])
edge_index = torch.stack([col, row], dim=0)
```
And as a result, the targets are the rows i.e. the centroids of the neighborhoods.